PhysicsTools/PatAlgos/src/SoftMuonMvaEstimator.cc

```diff
@@ -0,0 +1,101 @@
+#include "PhysicsTools/PatAlgos/interface/SoftMuonMvaEstimator.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+
+using namespace pat;
+
+SoftMuonMvaEstimator::SoftMuonMvaEstimator():
+	tmvaReader_("!Color:!Silent:Error"),
+	initialized_(false),
+	mva_(0)
+{}
+
+void SoftMuonMvaEstimator::initialize(std::string weightsfile)
+{
+	if (initialized_) return;
+	tmvaReader_.AddVariable("segComp",			&segmentCompatibility_);
+	tmvaReader_.AddVariable("chi2LocMom",			&chi2LocalMomentum_);
+	tmvaReader_.AddVariable("chi2LocPos",			&chi2LocalPosition_);
+	tmvaReader_.AddVariable("glbTrackTailProb",		&glbTrackProbability_);
+	tmvaReader_.AddVariable("iValFrac",			&iValidFraction_);
+	tmvaReader_.AddVariable("LWH",				&layersWithMeasurement_);
+	tmvaReader_.AddVariable("kinkFinder",			&trkKink_);
+	tmvaReader_.AddVariable("TMath::Log(2+glbKinkFinder)",	&log2PlusGlbKink_);
+	tmvaReader_.AddVariable("timeAtIpInOutErr",		&timeAtIpInOutErr_);
+	tmvaReader_.AddVariable("outerChi2",			&outerChi2_);
+	tmvaReader_.AddVariable("innerChi2",			&innerChi2_);
+	tmvaReader_.AddVariable("trkRelChi2",			&trkRelChi2_);
+	tmvaReader_.AddVariable("vMuonHitComb",			&vMuonHitComb_);
+	tmvaReader_.AddVariable("Qprod",			&Qprod_);
+
+	tmvaReader_.AddSpectator("pID",			       	&pID_);
+	tmvaReader_.AddSpectator("pt",				&pt_);
+	tmvaReader_.AddSpectator("eta",				&eta_);
+	tmvaReader_.AddSpectator("MomID",			&MomID_);
+
+	tmvaReader_.BookMVA("BDT", weightsfile);
+	initialized_ = true;	
+}
+
+void SoftMuonMvaEstimator::computeMva(const pat::Muon& muon)
+{
+	if (not initialized_)
+		throw cms::Exception("FatalError") << "SoftMuonMVA is not initialized";
+
+	reco::TrackRef gTrack = muon.globalTrack();
+	reco::TrackRef iTrack = muon.innerTrack();
+	reco::TrackRef oTrack = muon.outerTrack();
+
+	if(!(muon.innerTrack().isNonnull() and 
+	     muon.outerTrack().isNonnull() and 
+	     muon.globalTrack().isNonnull()))
+	{
+	  mva_ = -1; 
+	  return;
+	}
+
+	//VARIABLE EXTRACTION
+	pt_ = muon.pt();
+	eta_ = muon.eta();
+	dummy_ = -1;
+	MomID_ = dummy_;
+	pID_ = dummy_;
+
+
+	chi2LocalMomentum_ = muon.combinedQuality().chi2LocalMomentum;
+	chi2LocalPosition_ = muon.combinedQuality().chi2LocalPosition;
+	glbTrackProbability_ =  muon.combinedQuality().glbTrackProbability;
+	trkRelChi2_ =  muon.combinedQuality().trkRelChi2;
+
+	trkKink_ = muon.combinedQuality().trkKink;
+	log2PlusGlbKink_ =  TMath::Log(2+muon.combinedQuality().glbKink);
+	segmentCompatibility_ = muon.segmentCompatibility();
+
+	timeAtIpInOutErr_ = muon.time().timeAtIpInOutErr;
+
+	//TRACK RELATED VARIABLES
+	
+	iValidFraction_ =  iTrack->validFraction();
+	innerChi2_ =  iTrack->normalizedChi2();
+	layersWithMeasurement_ = static_cast<float>(iTrack->hitPattern().trackerLayersWithMeasurement());
+
+	outerChi2_ = oTrack->normalizedChi2();
+
+	Qprod_ =  static_cast<float>(iTrack->charge()) * static_cast<float>(oTrack->charge());
```

Comment:
was this casting really necessary?

Reply:
The TMVA method use all the variable as float. Since I don't know if TMVA has an in build casting function I made a explicit casting here. I don't think that removing it will cause any problem. Should I remove it?

Reply:
the left hand side of the expression (Qprod_) is declared as float and it is used in the TMVA inputs.
So, unless you have some evidence of bad action by compiler which somehow leads to ints passed to TMVA, I think that the casts are not necessary.

Reply:
I'll remove the casting in the next commit. I'll remove also casting on line 80 for "layersWithMeasurement_".